README.md
# Wunderman Start Project

Start project baseado em gulp e bower para instala&ccedil;&atilde;o de deped&ecirc;ncias.

### Detalhes da aplica&ccedil;&atilde;o
  - [Camada de back-end](sources/backend/)
  - [Camada de front-end](sources/frontend/)

### Vers&atilde;o
  - v1.0.0 (29/06/2016)

### [**Release notes**](release_notes.md)

release_notes.md
###Realease Notes
- [**v1.0.0**](#v1.0.0)
- [**v1.0.1**](#v1.0.1)
- [**v1.0.2**](#v1.0.2)
- [**v1.0.3**](#v1.0.3)
- [**v1.0.4**](#v1.0.4)

#v1.0.0
<a name="v1.0.0"></a>
	- Description

#v1.0.1
<a name="v1.0.1"></a>
	- Description

#v1.0.2
<a name="v1.0.2"></a>
	- Description

#v1.0.3
<a name="v1.0.3"></a>
	- Description

#v1.0.4
<a name="v1.0.4"></a>
	- Description

samples/InfiniteTubes/js/demo1.js
// Get window dimension
var ww = document.documentElement.clientWidth || document.body.clientWidth;
var wh = window.innerHeight;

// Save half window dimension
var ww2 = ww * 0.5, wh2 = wh * 0.5;

// Constructor function
function Tunnel() {
  // Init the scene and the
  this.init();
  // Create the shape of the tunnel
  this.createMesh();

  // Mouse events & window resize
  this.handleEvents();

  // Start loop animation
  window.requestAnimationFrame(this.render.bind(this));
}

Tunnel.prototype.init = function() {
  // Define the speed of the tunnel
  this.speed = 0.02;

  // Store the position of the mouse
  // Default is center of the screen
  this.mouse = {
    position: new THREE.Vector2(0, 0),
    target: new THREE.Vector2(0, 0)
  };

  // Create a WebGL renderer
  this.renderer = new THREE.WebGLRenderer({
    antialias: true,
    canvas: document.querySelector("#scene")
  });
  // Set size of the renderer and its background color
  this.renderer.setSize(ww, wh);
  this.renderer.setClearColor(0x222222);

  // Create a camera and move it along Z axis
  this.camera = new THREE.PerspectiveCamera(15, ww / wh, 0.01, 1000);
  this.camera.position.z = 0.35;

  // Create an empty scene and define a fog for it
  this.scene = new THREE.Scene();
  this.scene.fog = new THREE.Fog(0x222222, 0.6, 2.8);
};

Tunnel.prototype.createMesh = function() {
  // Empty array to store the points along the path
  var points = [];

  // Define points along Z axis to create a curve
  for (var i = 0; i < 5; i += 1) {
    points.push(new THREE.Vector3(0, 0, 2.5 * (i / 4)));
  }
  // Set custom Y position for the last point
  points[4].y = -0.06;

  // Create a curve based on the points
  this.curve = new THREE.CatmullRomCurve3(points);
  // Define the curve type

  // Empty geometry
  var geometry = new THREE.Geometry();
  // Create vertices based on the curve
  geometry.vertices = this.curve.getPoints(70);
  // Create a line from the points with a basic line material
  this.splineMesh = new THREE.Line(geometry, new THREE.LineBasicMaterial());

  // Create a material for the tunnel with a custom texture
  // Set side to BackSide since the camera is inside the tunnel
  this.tubeMaterial = new THREE.MeshStandardMaterial({
    side: THREE.BackSide,
    map: textures.stone.texture,
    bumpMap: textures.stoneBump.texture,
    bumpScale: 0.0003
  });
  
  // Add two lights in the scene
  // An hemisphere light, to add different light from sky and ground
  var light = new THREE.HemisphereLight( 0xffffbb, 0x887979, 0.9);
this.scene.add( light );
  // Add a directional light for the bump
  var directionalLight = new THREE.DirectionalLight( 0xffffff, 0.8 );
this.scene.add( directionalLight );
  // Repeat the pattern
  this.tubeMaterial.map.wrapS = THREE.RepeatWrapping;
  this.tubeMaterial.map.wrapT = THREE.RepeatWrapping;
  this.tubeMaterial.map.repeat.set(30, 6);
  this.tubeMaterial.bumpMap.wrapS = THREE.RepeatWrapping;
  this.tubeMaterial.bumpMap.wrapT = THREE.RepeatWrapping;
  this.tubeMaterial.bumpMap.repeat.set(30, 6);

  // Create a tube geometry based on the curve
  this.tubeGeometry = new THREE.TubeGeometry(this.curve, 70, 0.02, 50, false);
  // Create a mesh based on the tube geometry and its material
  this.tubeMesh = new THREE.Mesh(this.tubeGeometry, this.tubeMaterial);
  // Push the tube into the scene
  this.scene.add(this.tubeMesh);

  // Clone the original tube geometry
  // Because we will modify the visible one but we need to keep track of the original position of the vertices
  this.tubeGeometry_o = this.tubeGeometry.clone();
};

Tunnel.prototype.handleEvents = function() {
  // When user resize window
  window.addEventListener("resize", this.onResize.bind(this), false);
  // When user move the mouse
  document.body.addEventListener(
    "mousemove",
    this.onMouseMove.bind(this),
    false
  );
};

Tunnel.prototype.onResize = function() {
  // On resize, get new width & height of window
  ww = document.documentElement.clientWidth || document.body.clientWidth;
  wh = window.innerHeight;
  ww2 = ww * 0.5;
  wh2 = wh * 0.5;

  // Update camera aspect
  this.camera.aspect = ww / wh;
  // Reset aspect of the camera
  this.camera.updateProjectionMatrix();
  // Update size of the canvas
  this.renderer.setSize(ww, wh);
};

Tunnel.prototype.onMouseMove = function(e) {
  // Save mouse X & Y position
  this.mouse.target.x = (e.clientX - ww2) / ww2;
  this.mouse.target.y = (wh2 - e.clientY) / wh2;
};

Tunnel.prototype.updateCameraPosition = function() {
  // Update the mouse position with some lerp
  this.mouse.position.x += (this.mouse.target.x - this.mouse.position.x) / 30;
  this.mouse.position.y += (this.mouse.target.y - this.mouse.position.y) / 30;

  // Rotate Z & Y axis
  this.camera.rotation.z = this.mouse.position.x * 0.2;
  this.camera.rotation.y = Math.PI - this.mouse.position.x * 0.06;
  // Move a bit the camera horizontally & vertically
  this.camera.position.x = this.mouse.position.x * 0.015;
  this.camera.position.y = -this.mouse.position.y * 0.015;
};

Tunnel.prototype.updateMaterialOffset = function() {
  // Update the offset of the material
  this.tubeMaterial.map.offset.x += this.speed;
};

Tunnel.prototype.updateCurve = function() {
  var index = 0, vertice_o = null, vertice = null;
  // For each vertice of the tube, move it a bit based on the spline
  for (var i = 0, j = this.tubeGeometry.vertices.length; i < j; i += 1) {
    // Get the original tube vertice
    vertice_o = this.tubeGeometry_o.vertices[i];
    // Get the visible tube vertice
    vertice = this.tubeGeometry.vertices[i];
    // Calculate index of the vertice based on the Z axis
    // The tube is made of 50 rings of vertices
    index = Math.floor(i / 50);
    // Update tube vertice
    vertice.x +=
      (vertice_o.x + this.splineMesh.geometry.vertices[index].x - vertice.x) /
      10;
    vertice.y +=
      (vertice_o.y + this.splineMesh.geometry.vertices[index].y - vertice.y) /
      5;
  }
  // Warn ThreeJs that the points have changed
  this.tubeGeometry.verticesNeedUpdate = true;

  // Update the points along the curve base on mouse position
  this.curve.points[2].x = -this.mouse.position.x * 0.1;
  this.curve.points[4].x = -this.mouse.position.x * 0.1;
  this.curve.points[2].y = this.mouse.position.y * 0.1;

  // Warn ThreeJs that the spline has changed
  this.splineMesh.geometry.verticesNeedUpdate = true;
  this.splineMesh.geometry.vertices = this.curve.getPoints(70);
};

Tunnel.prototype.render = function() {
  // Update material offset
  this.updateMaterialOffset();

  // Update camera position & rotation
  this.updateCameraPosition();

  // Update the tunnel
  this.updateCurve();

  // render the scene
  this.renderer.render(this.scene, this.camera);

  // Animation loop
  window.requestAnimationFrame(this.render.bind(this));
};

// All needed textures
var textures = {
  "stone": {
    url: 'img/demo3/galaxyTexture.jpg'
    // url: "img/demo1/stonePattern.jpg"
  },
  "stoneBump": {
    url: "img/demo3/galaxyTexture.jpg"
    // url: "img/demo1/stonePatternBump.jpg"
  }
};
// Create a new loader
var loader = new THREE.TextureLoader();
// Prevent crossorigin issue
loader.crossOrigin = "Anonymous";
// Load all textures
for (var name in textures) {
  (function(name) {
  loader.load(textures[name].url, function(texture) {
    textures[name].texture = texture;
    checkTextures();
  });
})(name)
}
var texturesLoaded = 0;
function checkTextures() {
  texturesLoaded++;
  if (texturesLoaded === Object.keys(textures).length) {
    document.body.classList.remove("loading");
    // When all textures are loaded, init the scene
    window.tunnel = new Tunnel();
  }
}


samples/winter-rush-master/src/snow.js
/**
	Winter Rush Snow
	Handles falling snow, bars, sky
	by Felix Turner / @felixturner / www.airtight.cc
**/

var WRSnow = function() {

	var SNOW_COUNT = 400;
	var SNOW_EDGE = 100;
	var SNOW_TOP = 1600;
	var SNOW_BOTTOM = -300;
	var BAR_COUNT = 20;
	
	var windDir = 0;
	var windStrength = 0;
	var snowTime = 0;
	var snowGeometry;
	var bars;
	var barMaterial;
	var skyMaterial;

	var snoise = new ImprovedNoise();

	function init(){


		//make falling snow
		snowGeometry = new THREE.Geometry();

		var snowSprite = THREE.ImageUtils.loadTexture( "res/img/snow.png" );

		for ( i = 0; i < SNOW_COUNT; i ++ ) {

			var vertex = new THREE.Vector3();
			vertex.x = ATUtil.randomRange(-WRConfig.FLOOR_WIDTH/2,WRConfig.FLOOR_WIDTH/2);
			vertex.y = ATUtil.randomRange(SNOW_BOTTOM,SNOW_TOP);
			vertex.z = ATUtil.randomRange(-WRConfig.FLOOR_DEPTH/2,WRConfig.FLOOR_DEPTH/2);

			snowGeometry.vertices.push( vertex );

		}

		var snowMaterial = new THREE.PointCloudMaterial( { 
			size: 50, 
			sizeAttenuation: true, 
			map: snowSprite, 
			transparent: true ,
			blending: THREE.AdditiveBlending,
			depthTest: true,
			opacity:0.7,
			depthWrite:false
		} );

		var particles = new THREE.PointCloud( snowGeometry, snowMaterial );
		WRGame.getMoverGroup().add( particles );

		//STRIPS
		//add bars for at high speed

		barMaterial = new THREE.MeshBasicMaterial({
			color: 0x0FF66FF,
			blending: THREE.AdditiveBlending,
			depthTest: false,
			transparent: true,
			opacity:.6,
			sizeAttenuation: true,
			side: THREE.DoubleSide,
		});

		var width = 4;
		var spread =  1000;

		var barGeom = new THREE.PlaneGeometry(20,500,1,1);

		bars = [];

		for (i = 0; i < BAR_COUNT; i++) {

			var bar = new THREE.Mesh( barGeom, barMaterial );

			bar.scale.x = ATUtil.randomRange(0.2,2);
			bar.origYScale = ATUtil.randomRange(0.2,2);
			bar.scale.z = ATUtil.randomRange(0.2,2);

			WRGame.getMoverGroup().add( bar );

			bar.rotation.x = Math.PI/2;
			bar.rotation.y = Math.PI/2;

			bar.position.x = ATUtil.randomRange(-WRConfig.FLOOR_WIDTH/2,WRConfig.FLOOR_WIDTH/2);
			bar.position.y = ATUtil.randomRange(-300,600);
			bar.position.z = ATUtil.randomRange(-WRConfig.FLOOR_DEPTH/2,WRConfig.FLOOR_DEPTH/2);

			bars.push(bar);

		}

		//SKY
		var textureSky = THREE.ImageUtils.loadTexture( "res/img/xmas-sky.jpg" );
		skyMaterial = new THREE.MeshBasicMaterial( {
			map:textureSky,
			transparent:true,
			depthTest: true,
			fog:false
		} );

		var planeGeometry = new THREE.PlaneGeometry( 800, 300,1,1 );
		skyMesh = new THREE.Mesh( planeGeometry, skyMaterial );
		WRMain.getScene().add( skyMesh );			
		skyMesh.scale.x = skyMesh.scale.y = 15;
		skyMesh.position.z = -3600;
		skyMesh.position.y = 1500;

	}

	function shift(){


		for(  i = 0; i < SNOW_COUNT; i++) {

			var vert = snowGeometry.vertices[i];
			vert.z += WRConfig.MOVE_STEP;

			if (vert.z + WRGame.getMoverGroup().position.z > WRConfig.FLOOR_DEPTH/2){
				vert.z	-= WRConfig.FLOOR_DEPTH;
			}			 

		}
		snowGeometry.verticesNeedUpdate = true;

		for (i = 0; i < BAR_COUNT; i++) {
			var p = bars[i].position;
			p.z += WRConfig.MOVE_STEP;
			if (p.z + WRGame.getMoverGroup().position.z > WRConfig.FLOOR_DEPTH/2){
				p.z	-= WRConfig.FLOOR_DEPTH;
			}		
		}

	}


	function animate(){

		//global perlin wind
		snowTime += 0.001;
		windStrength = snoise.noise(snowTime,0,0)*20;
		windDir = (snoise.noise(snowTime + 100,0,0) + 1)/2 * Math.PI*2;

		for(  i = 0; i < SNOW_COUNT; i++) {
			var vert = snowGeometry.vertices[i];

			//gravity
			vert.y -= 3;

			//bounds wrapping
			if (vert.y < SNOW_BOTTOM){
				vert.y = SNOW_TOP;
			}

			//only do fancy wind if not playing
			if (!WRGame.getPlaying()){

				vert.x += Math.cos(windDir)*windStrength;
				// vert.z += Math.sin(windDir)*windStrength;

					//wrap around edges
				if (vert.x > WRConfig.FLOOR_WIDTH/2 + SNOW_EDGE) vert.x = -WRConfig.FLOOR_WIDTH/2 + SNOW_EDGE;
				if (vert.x < -WRConfig.FLOOR_WIDTH/2 + SNOW_EDGE) vert.x = WRConfig.FLOOR_WIDTH/2 + SNOW_EDGE;

				if (vert.z > WRConfig.FLOOR_DEPTH/2 + SNOW_EDGE) vert.z = -WRConfig.FLOOR_DEPTH/2 + SNOW_EDGE;
				if (vert.z < -WRConfig.FLOOR_DEPTH/2 + SNOW_EDGE) vert.z = WRConfig.FLOOR_DEPTH/2 + SNOW_EDGE;
		
			}

		}
		snowGeometry.verticesNeedUpdate = true;

		var opac = (WRGame.getSpeed() - 0.5) *2;

		barMaterial.opacity = opac*2/3;
		skyMaterial.opacity = opac;

		for (i = 0; i < BAR_COUNT; i++) {
			var p = bars[i].position;
			p.z +=40;

			bars[i].scale.y = bars[i].origYScale * opac;
		}


	}

	return {
		init:init,
		animate:animate,
		shift:shift
	};


}();

samples/endlessgame/main.js
var PLANE_WIDTH = 50,
    PLANE_LENGTH = 1000,
    PADDING = PLANE_WIDTH / 5 * 2,
    POWERUP_COUNT = 10;

var axishelper = {},
    camera = {},
		$container = {},
    controls = {},
    containerWidth = 0,
    containerHeight = 0,
    directionalLight = {},
    globalRenderID = {},
    hero = {},
    hemisphereLight = {},
    mountain = {},
    mountains = [],
    plane = {},
    planeGeometry = {},
    planeMaterial = {},
    powerup = {},	
		powerups = [],
    powerupSpawnIntervalID = {},
    powerupCounterIntervalID = {},
    queue = {},
    renderer = {},
    scene = {},
    sky = {},
    skyGeometry = {},
    skyMaterial = {},
    skyTexture = {};

function render () {
  globalRenderID = requestAnimationFrame( render );
  controls.update(); 
  
  powerups.forEach( function ( element, index ) {
    powerups[ index ].animate();
  });
  
  mountains.forEach( function ( element, index ) {
    mountains[ index ].animate();
  });
  
  if ( detectCollisions( powerups ) === true ) {
    gameOver();
  }
  
	renderer.render( scene, camera );
}

function gameOver () {
  cancelAnimationFrame( globalRenderID );
  window.clearInterval( powerupSpawnIntervalID );
  window.clearInterval( powerupCounterIntervalID );
  
  $( '#overlay-gameover' ).fadeIn( 100 );
  
  $( '#btn-restart' ).one( 'click', function () {
    $( '#overlay-gameover' ).fadeOut( 50 );
    POWERUP_COUNT = 10;
    powerups.forEach( function ( element, index ) {
      scene.remove( powerups[ index ] );
    }); 
    powerups = [];
    hero.position.x = 0;
    render();
    startPowerupLogic();
  } );
}

function onWindowResize () {
  containerWidth = $container.innerWidth();
	containerHeight = $container.innerHeight();
  camera.aspect = containerWidth / containerHeight;
	camera.updateProjectionMatrix();
  renderer.clear();
	renderer.setSize( containerWidth, containerHeight );
}

function detectCollisions( objects ) {
  var origin = hero.position.clone();

  for ( var v = 0, vMax = hero.geometry.vertices.length; v < vMax; v += 1 ) {       
    var localVertex = hero.geometry.vertices[ v ].clone();
    var globalVertex = localVertex.applyMatrix4( hero.matrix );
    var directionVector = globalVertex.sub( hero.position );

    var ray = new THREE.Raycaster( origin, directionVector.clone().normalize() );
    var intersections = ray.intersectObjects( objects );
    if ( intersections.length > 0 && 
        intersections[ 0 ].distance < directionVector.length() ) {
      return true;
    }
  }
  return false;
}  

function getRandomInteger( min, max ) {
  return Math.floor( Math.random() * ( max - min + 1 ) ) + min;
}

function Hero () {
  var hero = {},
      heroGeometry = {},
      heroMaterial = {};
  
  heroGeometry = new THREE.CylinderGeometry( 0, 2, 5, 10 );
  heroMaterial = new THREE.MeshLambertMaterial( {
    color: 0xE91E63,
    shading: THREE.FlatShading
  } );
  hero = new THREE.Mesh( heroGeometry, heroMaterial );
  hero.castShadow = true;
  hero.position.set( 0, 5, ( PLANE_LENGTH / 2 ) );
  hero.rotation.x = 0.785;
  
  window.addEventListener( 'keydown', function () {
    if ( event.keyCode === 37 && hero.position.x !== -( PLANE_WIDTH - PADDING ) / 2 ) {
      hero.position.x -= ( PLANE_WIDTH - PADDING ) / 2;
    } else if ( event.keyCode === 39 && hero.position.x !== ( PLANE_WIDTH - PADDING ) / 2 ) {
      hero.position.x += ( PLANE_WIDTH - PADDING ) / 2;
    }
  } );
  
  return hero;
}

function createLandscapeFloors () {
  var planeLeft = {},
      planeLeftGeometry = {},
      planeLeftMaterial = {},
      planeRight = {};
  
  planeLeftGeometry = new THREE.BoxGeometry( PLANE_WIDTH, PLANE_LENGTH + PLANE_LENGTH / 10, 1 );
  planeLeftMaterial = new THREE.MeshLambertMaterial( {
    color: 0x8BC34A
  } );
  planeLeft = new THREE.Mesh( planeLeftGeometry, planeLeftMaterial );
  planeLeft.receiveShadow = true;
	planeLeft.rotation.x = 1.570;
  planeLeft.position.x = -PLANE_WIDTH;
  planeLeft.position.y = 1;
  
  planeRight = planeLeft.clone();
  planeRight.position.x = PLANE_WIDTH;
  
  scene.add( planeLeft, planeRight );
}

function createMountain ( i, isEast ) {
  var loader = {},
      prototype = {},
      object = {},
      objectDimensionX = {},
      objectDimensionY = {},
      objectDimensionZ = {};
  
  loader = new THREE.ColladaLoader();
  
  function createObject () {
    object = prototype.clone();
    objectDimensionX = Math.random() * 0.25 + 0.05;
    objectDimensionY = Math.random() * 0.25;
    objectDimensionZ = objectDimensionX;
    object.scale.set( objectDimensionX, objectDimensionY, objectDimensionZ );
    
    if ( isEast === true ) {
      object.position.x = PLANE_WIDTH * 2
      object.position.z = ( i * PLANE_LENGTH / 27 ) - ( 1.5 * PLANE_LENGTH );
    } else {
      object.position.x = -PLANE_WIDTH * 2
      object.position.z = ( i * PLANE_LENGTH / 27 ) - ( PLANE_LENGTH / 2 );
    }
    
    object.visible = true;
    
    object.animate = function () {
      if ( object.position.z < PLANE_LENGTH / 2 - PLANE_LENGTH / 10 ) {
        object.position.z += 5;
      } else {
        object.position.z = -PLANE_LENGTH / 2;
      }
    }
    
    mountains.push( object );
    scene.add( object );
  }
  
  loader.load(
    'https://s3-us-west-2.amazonaws.com/s.cdpn.io/26757/mountain.dae',
    function ( collada ) {
      prototype = collada.scene;
      prototype.visible = false;
      createObject();
    } );
  
}

function createSpotlights () {
  var spotLight = {},
      target = {},
      targetGeometry = {},
      targetMaterial = {};
  for ( var i = 0; i < 5; i += 1 ) {
    targetGeometry = new THREE.BoxGeometry(1, 1, 1);
    targetMaterial = new THREE.MeshNormalMaterial();
    target = new THREE.Mesh( targetGeometry, targetMaterial );
    target.position.set( 0, 2, ( i * PLANE_LENGTH / 5 ) - ( PLANE_LENGTH / 2.5 ) );
    target.visible = false;
    scene.add( target );
    
    spotLight = new THREE.SpotLight( 0xFFFFFF, 2 );
    spotLight.position.set( 150, ( i * PLANE_LENGTH / 5 ) - ( PLANE_LENGTH / 2.5 ), -200 );
    spotLight.castShadow = true;
    spotLight.shadowCameraNear = 10;
    spotLight.shadowCameraVisible = false;
    spotLight.target = target;
    spotLight.shadowMapWidth = 2048;
    spotLight.shadowMapHeight = 2048;
    spotLight.fov = 40;
    
    plane.add( spotLight );
  }
}

function PowerUp () {
  var object = {},
      objectDimension = 0,
      objectGeometry = {},
      objectMaterial = {},
      xPosition = 0,
      xPositionValues = [],
      yPosition = 0,
      yPositionValues = [],
      zPosition = 0,
      zPositionValues = [];
  
  objectDimension = 2;
  
  xPositionValues = [ -( PLANE_WIDTH - PADDING ) / 2, 0, ( PLANE_WIDTH - PADDING ) / 2 ];
  yPositionValues = [ objectDimension + 1 ];
  zPositionValues = [ -( PLANE_LENGTH - PADDING ) / 2 ];
  
  xPosition = xPositionValues[ getRandomInteger( 0, xPositionValues.length - 1 ) ];
  yPosition = yPositionValues[ getRandomInteger( 0, yPositionValues.length - 1 ) ];
  zPosition = zPositionValues[ getRandomInteger( 0, zPositionValues.length - 1 ) ];
  
  objectGeometry = new THREE.BoxGeometry( objectDimension, objectDimension, objectDimension, objectDimension );
  objectMaterial = new THREE.MeshLambertMaterial( {
    color: 0x29B6F6,
  	shading: THREE.FlatShading
  } );
  object = new THREE.Mesh( objectGeometry, objectMaterial );
  object.position.set( xPosition, yPosition, zPosition );
  object.castShadow = true;
  object.receiveShadow = true;
  
  object.animate = function () {
    
    if ( object.position.z < PLANE_LENGTH / 2 + PLANE_LENGTH / 10 ) {
      object.position.z += 10;
    } else {
      object.position.x = xPositionValues[ getRandomInteger( 0, xPositionValues.length - 1 ) ];
      object.position.z = -PLANE_LENGTH / 2;
    }
    
  }
  
  return object;
}

function startPowerupLogic () {
  powerupSpawnIntervalID = window.setInterval( function () {
  
    if ( powerups.length < POWERUP_COUNT ) {
      powerup = new PowerUp();
      powerups.push( powerup );
      scene.add( powerup );
    }

  }, 4000 );

  powerupCounterIntervalID = window.setInterval( function () {
    POWERUP_COUNT += 1;
  }, 30000 );
}

function initGame () {
  THREE.ImageUtils.crossOrigin = '';

  $container = $( '#container' );
  containerWidth = $container.innerWidth();
  containerHeight = $container.innerHeight();

  renderer = new THREE.WebGLRenderer();
  renderer.setSize( containerWidth, containerHeight );
  renderer.antialias = true;
  renderer.setClearColor( 0xFFFFFF, 1 );
  renderer.shadowMapEnabled = true;
  renderer.shadowMapSoft = true;
  $container.get( 0 ).appendChild( renderer.domElement );

  scene = new THREE.Scene();

  axishelper = new THREE.AxisHelper( PLANE_LENGTH / 2 );


  /* CAMERA */
  camera = new THREE.PerspectiveCamera( 45, containerWidth / containerHeight, 1, 3000 );
  camera.position.set( 0, PLANE_LENGTH / 125, PLANE_LENGTH / 2 + PLANE_LENGTH / 25 );


  /* CONTROLS */
  controls = new THREE.OrbitControls( camera, $container.get( 0 ) );
  controls.noKeys = true;
  controls.noPan = true;
  controls.noZoom = true; 
  controls.minPolarAngle = 1.55;
  controls.maxPolarAngle = 1.55;
  controls.minAzimuthAngle = 0;
  controls.maxAzimuthAngle = 0;


  /* FLOOR */
  planeGeometry = new THREE.BoxGeometry( PLANE_WIDTH, PLANE_LENGTH + PLANE_LENGTH / 10, 1 );
  planeMaterial = new THREE.MeshLambertMaterial( {
    color: 0x78909C
  } );
  plane = new THREE.Mesh( planeGeometry, planeMaterial );
  plane.rotation.x = 1.570;
  plane.receiveShadow = true;



  /* LANDSCAPE */
  createLandscapeFloors();
  for ( var i = 0; i < 60; i += 1 ) {
    var isEast = false;
    if ( i > 29 ) {
      isEast = true;
    }
    createMountain( i, isEast );
  }

  skyGeometry = new THREE.BoxGeometry( 1200, 800, 1, 1 );
  skyMaterial = new THREE.MeshBasicMaterial( {
    map: THREE.ImageUtils.loadTexture( 'https://s3-us-west-2.amazonaws.com/s.cdpn.io/26757/background.jpg' ),
    depthWrite: false,
    side: THREE.BackSide
  } );
  sky = new THREE.Mesh( skyGeometry, skyMaterial );
  sky.position.y = 300;
  sky.position.z = -PLANE_LENGTH / 2 + PADDING;


  /* LIGHTS */
  createSpotlights();
  directionalLight = new THREE.DirectionalLight( 0xffffff, 1 );
  directionalLight.position.set( 0, 1, 0 );
  hemisphereLight = new THREE.HemisphereLight( 0xFFB74D, 0x37474F, 1 );
  hemisphereLight.position.y = 500;


  /* POWERUPS */
  startPowerupLogic();


  /* HERO */
  hero = new Hero();


  /* SCENE */
  scene.add( camera, directionalLight, hemisphereLight, plane, sky, hero );
}

function runGame () {
  window.addEventListener( 'resize', onWindowResize );
  render();
  onWindowResize();
}

initGame();
runGame();
